### PR TITLE
[NU-1921] Add median aggregator

### PR DESCRIPTION
**Pull request by paw787878**
## Checklist before merge
- [x] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [x] Code is cleaned from temporary changes and commented out lines
- [x] Parts of the code that are not easy to understand are documented in the code
- [x] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [x] Documentation added or updated
- [x] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [x] Verify that PR will be squashed during merge

### CODE REVIEW

File: docs/scenarios_authoring/AggregatesInTimeWindows.md [hunk 1]
```diff
@@ -85,6 +85,7 @@ Let’s map the above statement on the parameters of the Nussknacker Aggregate c
 * StddevSamp - computes sample standard deviation
 * VarPop - computes population variance
 * VarSamp - computes sample variance
+* Median - computes median
```
Comment (jedrz) [inline docs/scenarios_authoring/AggregatesInTimeWindows.md:88]: Add an example in the table below?

File: engine/flink/components/base-tests/src/test/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/AggregatesSpec.scala [hunk 2]
```diff
@@ -154,6 +155,77 @@ class AggregatesSpec extends AnyFunSuite with TableDrivenPropertyChecks with Mat
...
+
+  test("should calculate correct results for median aggregator on BigInt") {
+    val agg = MedianAggregator
+    resultOnElements(List(new BigInteger("7"), new BigInteger("8")), agg) shouldEqual new java.math.BigDecimal("7.5")
+  }
+
+  test("should calculate correct results for median aggregator on floats") {
+    val agg    = MedianAggregator
+    val result = resultOnElements(List(7.0f, 8.0f), agg)
+    require(result.isInstanceOf[Double])
```
Comment (jedrz) [inline engine/flink/components/base-tests/src/test/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/AggregatesSpec.scala]: Use scalatest matcher for checking a type of variable

File: utils/math-utils/src/main/scala/pl/touk/nussknacker/engine/util/MathUtils.scala [hunk 3]
```diff
@@ -183,7 +183,7 @@ trait MathUtils {
     case n1: java.math.BigDecimal => n1.negate()
   }
 
-  private def compare(n1: Number, n2: Number): Int = {
```
Comment (jedrz) [inline utils/math-utils/src/main/scala/pl/touk/nussknacker/engine/util/MathUtils.scala:189]: MathUtils is exposed as a helper to a user. I think those public methods should not be used in scenarios. Annotate them with `@Hidden`?

File: engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/median/MedianHelper.scala [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import pl.touk.nussknacker.engine.util.MathUtils
+
+import javax.annotation.Nullable
+import scala.annotation.tailrec
+import scala.util.Random
+
+object MedianHelper {
+  private val rand = new Random(42)
+
+  @Nullable
```
Comment (jedrz) [inline engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/median/MedianHelper.scala]: Return Option here and convert to null  at the very end in the aggregate helper?

File: engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/median/MedianHelper.scala [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  @tailrec
+  private def quickSelect(numbers: List[Number], indexToTake: Int): Number = {
+    require(numbers.nonEmpty)
+    require(indexToTake >= 0)
+    require(indexToTake < numbers.size)
+
+    val randomElement = numbers(rand.nextInt(numbers.size))
+    val groupedBy = numbers.groupBy(e => {
+      val cmp = MathUtils.compare(e, randomElement)
+      if (cmp < 0) {
```
Comment (jedrz) [inline engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/median/MedianHelper.scala:39]: Is the if-else needed? I think `MathUtils.compare` should return -1, 0 or 1?

File: engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/median/MedianHelper.scala [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      val firstNumber  = quickSelect(numbers, numbers.size / 2 - 1)
+      val secondNumber = quickSelect(numbers, numbers.size / 2)
+
+      val sum = MathUtils.largeFloatingSum(firstNumber, secondNumber)
+      MathUtils.divideWithDefaultBigDecimalScale(sum, 2)
+    }
+  }
+
+  @tailrec
+  private def quickSelect(numbers: List[Number], indexToTake: Int): Number = {
```
Comment (jedrz) [inline engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/median/MedianHelper.scala:31]: Add a link to descirption of the algorithm, e.g. a Wikipedia link?

File: engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/median/MedianHelper.scala [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    val randomElement = numbers(rand.nextInt(numbers.size))
+    val groupedBy = numbers.groupBy(e => {
+      val cmp = MathUtils.compare(e, randomElement)
+      if (cmp < 0) {
+        -1
+      } else if (cmp == 0) {
+        0
+      } else 1
+    })
+    val smallerNumbers = groupedBy.getOrElse(-1, List())
```
Comment (jedrz) [inline engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/median/MedianHelper.scala]: You can use Nil as an empty list.

File: engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/aggregates.scala [hunk 8]
```diff
@@ -69,19 +74,38 @@ object aggregates {
...
+    override def computeStoredType(input: TypingResult): Validated[String, TypingResult] = Valid(
+      Typed.genericTypeClass[ListBuffer[_]](List(input))
+    )
+
+  }
+
   object ListAggregator extends Aggregator {
 
-    override type Aggregate = List[AnyRef]
+    override type Aggregate = ListBuffer[AnyRef]
```
Comment (jedrz) [inline engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/aggregates.scala]: I'm afraid it can break state compatibility in Flink jobs. It would be more safe to extract these changes to a separate PR and discuss it.

File: docs/scenarios_authoring/AggregatesInTimeWindows.md (hunk 1 above)
Comment (paw787878) [reply docs/scenarios_authoring/AggregatesInTimeWindows.md:88]: as I wrote in previous pr I think that more examples there is not needed

File: engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/median/MedianHelper.scala (hunk 5 above)
Comment (paw787878) [reply engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/median/MedianHelper.scala:39]: in the comparable interface compareTo, (I assume that this method has similar contract), it is documented that it returns negative if first is lower. So it can return value other than those three values. Should I move this if to the MathUtils.compare implementation?

Comment (jedrz) [reply engine/flink/components/base-unbounded/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/median/MedianHelper.scala:39]: You're right. There is Math.signum but it is defined only for float and double.

File: docs/Changelog.md [hunk 9]
```diff
@@ -27,6 +27,7 @@
 * [#7184](https://github.com/TouK/nussknacker/pull/7184) Improve Nu Designer API notifications endpoint, to include events related to currently displayed scenario
 * [#7323](https://github.com/TouK/nussknacker/pull/7323) Improve Periodic DeploymentManager db queries
 * [#7307](https://github.com/TouK/nussknacker/pull/7307) Added StddevPop, StddevSamp, VarPop and VarSamp aggregations
+* [#7321](https://github.com/TouK/nussknacker/pull/7321) Added Median aggregation
```
Comment (jedrz) [inline docs/Changelog.md]: aggregation -> aggregator

**Review by jedrz — APPROVED** [review]

File: engine/flink/components/base-tests/src/test/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/TransformersTest.scala [hunk 10]
```diff
@@ -455,6 +475,19 @@ class TransformersTest extends AnyFunSuite with FlinkSpec with Matchers with Ins
...
+    val id = "1"
+
+    val model =
+      modelData(List(TestRecordHours(id, 0, 1, "a")))
+    val testProcess = tumbling("#AGG.median", "#input.eId", emitWhen = TumblingWindowTrigger.OnEndWithExtraWindow)
+
+    val aggregateVariables = runCollectOutputAggregate[Number](id, model, testProcess)
+    aggregateVariables.length shouldEqual (2)
+    aggregateVariables(0) shouldEqual 1.0
+    require((aggregateVariables(1).asInstanceOf[Double].isNaN))
```
Comment (jedrz) [inline engine/flink/components/base-tests/src/test/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/TransformersTest.scala]: Use matcher :)
